Use a serial line buffer to prevent queue corruption

thinkyhead opened the pull request:
In addition to getting rid of "ok" messages for all non-serial commands...

This PR takes another approach to fix the issue where `enqueuecommand` causes the command queue to become corrupted. To sum up the bug, if any code uses `enqueuecommand` or `enqueuecommands_P` while there's a serial stream (as with an active print job), if there is an incomplete command in the buffer, the injected command will first overwrite the current buffer slot, then the next buffer slot will get only a partial serial command written into it, causing unpredictable results due to stale data in the ring buffer.

This approach adds an intermediate buffer to cache a single line of serial input. This ensures that only complete commands are ever injected into the queue. So there's no incomplete writing state, and calls to `enqueuecommand` can't corrupt some serial command in-progress.

As a result, code doesn't need to worry about whether `serial_count` is zero or not, and the SD buffer continues to process in the middle of an incomplete serial command.
- To prevent spurious re-running of dead commands in the queue, one additional trick would be to write `null` to the first byte of the command slot immediately after processing it.

References: #2911 #2951 #2984

thinkyhead committed:
Use a separate serial line buffer

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -946,24 +954,25 @@ void get_command() {
-        }
-        command_queue[cmd_queue_index_w][serial_count] = 0; //terminate string
-        // if (!comment_mode) {
+        if (!sd_count) continue; //skip empty lines
```
Note here that empty lines `continue` instead of `return`ing.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -896,44 +893,55 @@ void get_command() {
-          serial_count >= (MAX_CMD_SIZE - 1) || n == -1
+      card_eof = card.eof();
+      char sd_char = (char)n;
+      if (card_eof || n == -1
```
Can end of file happen without LF or CR? If so this extra condition will get the last line.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -792,9 +789,12 @@ void gcode_line_error(const char* err, bool doFlush = true) {
   if (drain_queued_commands_P()) return; // priority is given to non-serial commands
 
-  #if ENABLED(NO_TIMEOUTS)
+  #if defined(NO_TIMEOUTS) && NO_TIMEOUTS > 0
```
Since `NO_TIMEOUTS` is set as an integer value we should be using `defined` here.

AnHardt commented:
Looks ok to me - but untested.

In a next step
- the line-ring-buffer could be a struct (line,line_number,send_ok) (number only used by ADVANCED_OK)
- collecting chars from the serial line into serial_line_buffer could be a function.
- we could store a 'cleaned' (without Nlinenumber and *checksum) line in the ring-buffer. That would prevent us from messing with the line-number in process_next_command() again. Line-number and checksum are property's of the serial protocol - not part of the g-code syntax.

Notes to myself:  
"A line with checksum can't contain a comment!" 
"Can the checksum be calculated on the fly - with reading from the serial.read()?"
"strtol() gives back the pointer to the position right behind the last read digit in the second parameter."
"strtol() works with spaces ahead."
"Searching for the '*' should be done with strrchr() to find the most right one. If then the pointer after strtol is [NULL|/n|/r|' '] and strtol gives a number, we have a checksum

thinkyhead commented:
All good ideas. And right, it is untested. I should give the line buffer a proper run and make sure it functions under normal conditions before calling it a bug fix.

It is possible to add extra logic to the line buffering, to do things like drop consecutive spaces. And while presumptuous, we could get around the problem of comments preventing the parsing of "*" checksums by assuming that any instance of asterisk-number a checksum —so the last will be used— but ignore it if no N was given.

AnHardt commented:
"Checksum without N" was a relative frequent error during my tests. Don't drop it.
Printing the malformed string in `gcode_line_error()` really helped debugging.

AnHardt commented:
No test possible. RH and Pronterface do crash.

AnHardt commented:
Added some of my debug otput but still cant see what's going on.

```
OLD DEBUG OUTPUT
```

That could be one of the serial line errors. 

> echo:enqueueing "B@:0 @:035" 

!!! Thats a partial M105 output.

```
DEBUG OUTPUT SHOWING "Ring buffer full" messages.
```

RH keeps sending but no line is consumed after N2.

```
DEBUG OUTPUT SHOWING "Ring buffer full" messages, no ok message after N2.
```

No oks after N2

```
DEBUG OUTPUT showing weird output
```

Pulled the USB cable.

AnHardt commented:
enqueuecommand() uses strcpy(), so we have to terminate the serial_line_buffer with a NULL.

thinkyhead commented:
@AnHardt Maybe my `static` is not being respected? It sounds entirely like a buffer corruption issue. Other data writing into the same RAM. I don't think it's a stack collision.

> do things like drop consecutive spaces

Obviously a bad idea that I retract. It would screw up the checksum.

> Pulled the USB cable.

Clearly I need to review the code and look for stupid errors.

thinkyhead commented:
@AnHardt I see one obvious error. `gcode_line_error` wants to reset the serial buffer and get a new line. So the original code sets `serial_count` to zero in that function. The new code fails to do that. So let me patch that so you can give it another test…

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -733,7 +732,7 @@ void setup() {
  */
 void loop() {
-  if (commands_in_queue < BUFSIZE - 1) get_command();
+  if (commands_in_queue < BUFSIZE) get_command();
```
`commands_in_queue` is 1-based.

thinkyhead commented:
Ok, this is pretty complete now and should be testable. If it passes inspection, perhaps it can go in as a bug patch and give us a simplified less blocking command queue. If this needs extra time for testing, then I can take the changes replacing "fromsd" with "send_ok" and make a small PR — since that should definitely be patched for the next RC.

jbrazio commented:
Why isn't this going into RC4 if the patch is ready ?
What kind of testing shall we perform ?

AnHardt commented on Marlin/Marlin_main.cpp:
```diff
@@ -903,44 +894,55 @@ void get_command() {
+      if (card_eof || n == -1
+          || sd_char == '\n' || sd_char == '\r' ||
+          || ((sd_char == '#' || sd_char == ':') && !comment_mode)
       ) {
```
with SDSUPPORT

```
sketch\Marlin_main.cpp: In function 'void get_command()':

Marlin_main.cpp:943: error: expected primary-expression before '||' token

           || ((sd_char == '#' || sd_char == ':') && !comment_mode)

           ^

exit status 1
expected primary-expression before '||' token
```

AnHardt commented:
Without the ecqueuing messages it would look a bit nicer.
Maybe with "ok"s it could work.
You are a good programmer, but a lousy tester.

```
01:37:02.284 : Printer reset detected - initalizing
01:37:02.300 : start
01:37:02.300 : echo: External Reset
01:37:02.300 : Marlin 1.1.0-RC3
01:37:02.300 : echo: Last Updated: 2015-12-01 12:00 | Author: (none, default config)
01:37:02.300 : Compiled: Mar 10 2016
01:37:02.300 : echo: Free Memory: 3722  PlannerBufferBytes: 1232
01:37:02.300 : echo:Hardcoded Default Settings Loaded
01:37:02.300 : echo:Steps per unit:
01:37:02.300 : echo:  M92 X80.00 Y80.00 Z4000.00 E500.00
01:37:02.300 : echo:Maximum feedrates (mm/s):
01:37:02.300 : echo:  M203 X300.00 Y300.00 Z5.00 E25.00
01:37:02.300 : echo:Maximum Acceleration (mm/s2):
01:37:02.300 : echo:  M201 X3000 Y3000 Z100 E10000
01:37:02.300 : echo:Accelerations: P=printing, R=retract and T=travel
01:37:02.303 : echo:  M204 P3000.00 R3000.00 T3000.00
01:37:02.310 : echo:Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)
01:37:02.313 : echo:  M205 S0.00 T0.00 B20000 X20.00 Z0.40 E5.00
01:37:02.314 : echo:Home offset (mm):
01:37:02.314 : echo:  M206 X0.00 Y0.00 Z0.00
01:37:02.315 : echo:Material heatup parameters:
01:37:02.316 : echo:  M145 S0 H180 B70 F0
01:37:02.319 : echo:  M145 S1 H240 B110 F0
01:37:02.319 : echo:PID settings:
01:37:02.321 : echo:  M301 P22.20 I1.08 D114.00 C100.00 L20
01:37:02.323 : echo:Filament settings: Disabled
01:37:02.324 : echo:  M200 D3.00
01:37:02.325 : echo:  M200 D0
01:37:02.510 : N1 M110*34
01:37:02.510 : N2 M115*36
01:37:02.510 : N4 M114*35
01:37:02.510 : N5 M111 S6*98
01:37:02.520 : N6 T0*60
01:37:02.520 : N7 M20*22
01:37:02.521 : N8 M80*19
01:37:05.668 : echo:enqueueing "N1 M110*34"
01:37:05.679 : echo:SD card ok
01:37:05.684 : echo:enqueueing "N2 M115*36"
01:37:05.732 : echo:SD card ok
01:37:05.763 : FIRMWARE_NAME:Marlin 1.1.0-RC3 From Archive SOURCE_CODE_URL:https://github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:3D Printer EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000
01:37:05.763 : echo:enqueueing "N4 M114*35"
01:37:05.764 : echo:enqueueing "N5 M111 S6*98"
01:37:05.778 : N11 M20*33
01:37:05.778 : echo:enqueueing "N6 T0*60"
01:37:05.779 : N12 M20*34
01:37:05.779 : N13 M220 S100*83
01:37:05.779 : N14 M221 S100*85
01:37:05.779 : N15 M111 S6*83
01:37:05.781 : N16 T0*13
01:38:41.522 : N41 M111 S14*97
01:38:43.916 : N43 M111 S15*98
01:39:09.293 : ????
```

RH crashed with bluescreen some minutes later.

AnHardt commented:
Test without SDSUPPORT was a tiny bit better but also without execution of any command and no ok's.
M105 seems to work - i see a graph. But i can't disconnect or reset the mega, nor close RH.

AnHardt commented:
Proterface hangs immediately after connect, but i see millions of M105 answers and can't disconnect or reset the Mega nor close Pronterface.

thinkyhead commented:
> You are a good programmer, but a lousy tester.

That's what I need everyone else for!

I'll keep debugging. And I'll try putting this onto my machine and testing it. Sorry I keep missing things. It's a hassle to flash the board, so I tend to avoid it.

AnHardt commented:
Most tests done here are made with a spare Mega / RAMPS / display / 2-thermistors combination without an attached machine. Put that on your wish-list and occasionally remind people you have one (with a link).

thinkyhead commented:
> Put that on your wish-list

@AnHardt It's a growing list. Although I haven't gotten those equipment donations, I've gotten some cash donations, helpfully covering some of my coffee costs. Because all any programmer really needs is caffeine.

So I've posted some more updates to this PR to cover potential issues.
- Fixed the typo in `|| sd_char == '\n' || sd_char == '\r'` …
- When the `serial_count` goes past the buffer length I'm restraining it so it can't point past the end of the buffer. This will keep a very long command string from causing `nul` to be written into unknown RAM (when the command concludes).
- I got rid of the echo on serial commands by splitting up `enqueuecommand` into two functions.
- This patch also groups code to do `M23`, `M24` because the LCD "SD Print" and `cardreader::autostart` routines had redundant code.
- Put the code that eats up a buffer item back into `loop()` because `M29` never calls `process_next_command`.
- Renamed `enqueuecommand`, `enqueuecommands_P` to `enqueue_and_echo_command`, `enqueue_and_echo_commands_P` for clarity.
- Separate `comment_mode` variables were needed for SD vs serial input. (Longstanding bug actually!)

I have just started testing this on my machine, so I might have further patches soon. You should probably wait until I give the "all clear" because there are bound to be glitches still.

thinkyhead commented:
Okay! So this now tests positively on my machine. Apparently the `comment_mode` variable was shared between serial input and sd input, and this was a potential problem. Splitting these up ensures that characters won't be ignored because some other input received a semicolon.

So, overall I think this gives us a more robust set of input buffers.

As for calculating the checksum as the characters arrive, the two buffers more or less preclude that. It's simpler (and may even be faster) to just calculate the checksum for each line when the commands are processed rather than as characters arrive.

thinkyhead commented:
> Why isn't this going into RC4 if the patch is ready? What kind of testing shall we perform?

It might be ready now that I've given it a thorough scrubbing. It surely wasn't ready before.

@jbrazio So to test this try doing the usual things that you do every day. Send commands from your host, try SD printing, try sending immediate commands from the host _during_ SD printing, etc. Basically, everything should now just work as before, except immediate commands will no longer interrupt or corrupt SD commands.

jbrazio commented:
Compiling and uploading right now..

```
Sketch uses 133,652 bytes (52%) of program storage space. Maximum is 253,952 bytes.
Global variables use 5,061 bytes (61%) of dynamic memory, leaving 3,131 bytes for local variables. Maximum is 8,192 bytes.
```

jbrazio commented:
Not looking good..
If I issue a command at a time like G28 no problems, when I tried printing from Prointerface:

```
Loading file: C:\Users\jbrazio\Desktop\test.gcode
Loaded C:\Users\jbrazio\Desktop\test.gcode, 1550 lines
27.25mm of filament used in this print
The print goes:
- from 4.65 mm to 43.25 mm in X and is 38.60 mm wide
- from 3.10 mm to 174.51 mm in Y and is 171.41 mm deep
- from 0.00 mm to 0.18 mm in Z and is 0.18 mm high
Estimated duration: 1 layers, 0:00:10
SENT: M105
RECV: ok T:53.8 /0.0 @:0
SENT: M105
RECV: ok T:52.8 /0.0 @:0
SENT: M105
RECV: ok T:51.8 /0.0 @:0
SENT: N-1 M110*15
Print started at: 03:29:29

[Does nothing]
```

Seem that the blocking command is `N-1 M110*15`.

thinkyhead commented:
Durn. What could it be? Try adding some helpful serial output to see what's going on.

After the line, around 826:

``` cpp
    char serial_char = MYSERIAL.read();
```

…add…

``` cpp
    SERIAL_ECHOPAIR("Received char ", (unsigned long)serial_char);
    SERIAL_ECHOPAIR(" - serial_count: ", (unsigned long)serial_count);
    SERIAL_EOL;
```

After the line, around 840:

``` cpp
      char* command = serial_line_buffer;
```

…add…

``` cpp
      SERIAL_ECHOPGM("Full line: ");
      SERIAL_ECHOLN(serial_line_buffer);
```

And after the line:

``` cpp
        serial_line_buffer[serial_count++] = serial_char;
```

…add…

``` cpp
        SERIAL_ECHOLNPGM("Storing char");
```

I'm testing the same on my end. Hopefully it reveals something. I haven't tried a host-based GCode session yet, but will mess with that more in the next round of testing.

jbrazio commented:
Here is the output.
I had to replace `SERIAL_ECHOLNPGM(" - serial_count: ", (unsigned long)serial_count);`with `SERIAL_ECHOPAIR(" - serial_count: ", (unsigned long)serial_count);`.

```
Printer is now online.
Loading file: C:\Users\jbrazio\Desktop\test.gcode
Loaded C:\Users\jbrazio\Desktop\test.gcode, 1550 lines
27.25mm of filament used in this print
The print goes:
- from 4.65 mm to 43.25 mm in X and is 38.60 mm wide
- from 3.10 mm to 174.51 mm in Y and is 171.41 mm deep
- from 0.00 mm to 0.18 mm in Z and is 0.18 mm high
Estimated duration: 1 layers, 0:00:10
Received char 77 - serial_count: 0
Received char 49 - serial_count: 1
Received char 48 - serial_count: 2
Received char 53 - serial_count: 3
Received char 10 - serial_count: 4
Full line: M105
Received char 77 - serial_count: 0
Received char 49 - serial_count: 1
Received char 48 - serial_count: 2
Received char 53 - serial_count: 3
Received char 10 - serial_count: 4
Full line: M105
Received char 77 - serial_count: 0
Received char 49 - serial_count: 1
Received char 48 - serial_count: 2
Received char 53 - serial_count: 3
Received char 10 - serial_count: 4
Full line: M105
Received char 78 - serial_count: 0
Received char 45 - serial_count: 1
Received char 49 - serial_count: 2
Received char 32 - serial_count: 3
Received char 77 - serial_count: 4
Received char 49 - serial_count: 5
Received char 49 - serial_count: 6
Received char 48 - serial_count: 7
Received char 42 - serial_count: 8
Received char 49 - serial_count: 9
Received char 53 - serial_count: 10
Print started at: 15:58:09
Received char 10 - serial_count: 11
Full line: N-1 M110*15

[Does nothing]
```

**-EDIT-**

I've been doing more debug and it seems `N-1 M110*15` gets into `enqueuecommand()` just fine.

```
Full line: N-1 M110*15
if (npos)
if (M110)
if (apos)
continue parsing
after  if (npos)
enqueuecommand(serial_line_buffer): true
```

thinkyhead commented:
@jbrazio Interesting. I will have to look now at `_enqueuecommand` and beyond to see why it doesn't like that line.

AnHardt commented:
Anyone seen an 'ok' so far?

jbrazio commented:
@AnHardt you're right, the `N-1 M110*15` command is never ack from Marlin, that maybe the reason why Pronterface hangs.

AnHardt commented:
```
inline bool _enqueuecommand(const char* cmd) {
  if (*cmd == ';' || commands_in_queue >= BUFSIZE) return false;
  char* command = command_queue[cmd_queue_index_w];
  strcpy(command, cmd);
  send_ok[cmd_queue_index_w] = false; //!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  cmd_queue_index_w = (cmd_queue_index_w + 1) % BUFSIZE;
  commands_in_queue++;
  return true;
}
```

jbrazio commented:
What' the reason behind the `send_ok` array ?
To only send "ok" replies to certain commands ?

AnHardt commented:
Its running in parallel to `command_queue[]` its the place where we remembe if to send a 'ok' or not.
When we set it to false always, we get never an 'ok'. Only for serial/USB input this should be true.

We set in line 843 but reset it in _enqueuecommand()

The simplest way is calling _enqueuecommand() with a additional parameter.

jbrazio commented:
So in a nutshell the sole propose for it to distinguish between SD vs. serial printing ?

AnHardt commented:
To distinguish serial from sd-printing and directly injected commands.

jbrazio commented:
What is considered a "directly injected command" ?

-edit-
Sorry for all these questions. :-)

AnHardt commented:
`enqueue_and_echo_commands_P()` formerly known as `enqueuecommands_P()`

AnHardt commented:
I see 'ok's

```
@@ -500,25 +500,25 @@ void enqueue_and_echo_commands_P(const char* pgcode) {

 /**
  * Copy a command directly into the main command buffer, from RAM.
  * Returns true if successfully adds the command
  */
-inline bool _enqueuecommand(const char* cmd) {
+inline bool _enqueuecommand(const char* cmd, bool okIt = false) {
   if (*cmd == ';' || commands_in_queue >= BUFSIZE) return false;
   char* command = command_queue[cmd_queue_index_w];
   strcpy(command, cmd);
-  send_ok[cmd_queue_index_w] = false;
+  send_ok[cmd_queue_index_w] = okIt;
   cmd_queue_index_w = (cmd_queue_index_w + 1) % BUFSIZE;
   commands_in_queue++;
   return true;
 }

 /**
  * Enqueue with Serial Echo
  */
 bool enqueue_and_echo_command(const char* cmd) {
-  if (_enqueuecommand(cmd)) {
+  if (_enqueuecommand(cmd), false) {
     SERIAL_ECHO_START;
     SERIAL_ECHOPGM(MSG_Enqueueing);
     SERIAL_ECHO(cmd);
     SERIAL_ECHOLNPGM("\"");
     return true;
@@ -837,13 +837,10 @@ void get_command() {
       serial_line_buffer[serial_count] = 0; // terminate string
       serial_count = 0; //reset buffer

       char* command = serial_line_buffer;

-      // This item in the queue is from the serial line
-      send_ok[cmd_queue_index_w] = true;
-
       while (*command == ' ') command++; // skip any leading spaces
       char* npos = (*command == 'N') ? command : NULL; // Require the N parameter to start the line
       char* apos = strchr(command, '*');

       if (npos) {
@@ -908,11 +905,11 @@ void get_command() {
       #if defined(NO_TIMEOUTS) && NO_TIMEOUTS > 0
         last_command_time = ms;
       #endif

       // Add the command to the queue
-      _enqueuecommand(serial_line_buffer);
+      _enqueuecommand(serial_line_buffer, true);
     }
     else if (serial_count >= MAX_CMD_SIZE - 1) {
       // Keep fetching, but ignore normal characters beyond the max length
       // The command will be injected when EOL is reached
     }
```

seems to run now.

AnHardt commented:
Still not ok.
Commands injected by the menu are repeated endlessly.
Starting a sd-file results in an endless "echo:M24"

thinkyhead commented:
Ok, I fixed the problem where "ok" wasn't being produced by serial commands. (Pronterface hides "ok" replies so this wasn't easily noticed.) I am now able to do host printing. I am about to try the same thing with SD printing, and will post my results.

**[edit]**

SD printing and direct command injection look much better now with these latest fixes. Blank lines and `sd_comment_mode` were both causing a SNAFU in the sd character processing.

I think you'll find that this is pretty solid now. But of course it needs more testing just to be sure!

AnHardt commented:
Works pretty good now. Much better than RC. No truncated lines in the buffer. No additional ok's.

Testfile was:

```
M111 S15
G28
G29
G29
G29
G28
```

Configured:
#define Z_PROBE_END_SCRIPT "G1 Z10 F12000\nG1 X15 Y130\nG1 Z0.5\nG1 Z10"

Tested with Pronterface, Octoprint, RH in PingPong-mode, RH without PingPong-mode.

Printing over usb
Ok for all hosts, but problems with out of sequence commands when enqueuing G29 with end-script (only RH without PingPong). That was completely expected and is much better than before where we corrupted the buffer.

File upload to sd
Is ok for all hosts but Octoprint 1.2,9. That's iss 1224 in Octoprint - not Marlins fault (some more gray hairs).

File delete from sd
Ok for all hosts.

File print from sd-card.
OK. But consecutive endscripts can overwrite each other because the next line of code from the sd is executed before the second line of the enqueued command. That can replace the the string to enqueue. The sequence of execution is messed up.

Let me repeat. That's much better than RC. My #2984 has **exactly** the same problems as i tried to express in https://github.com/MarlinFirmware/Marlin/pull/2984#discussion_r53309084.

The solution could be, to block any new commands, from any other source, until the last line in the sequence ( "G1 Z10 F12000\nG1 X15 Y130\nG1 Z0.5\nG1 Z10") is queued in.

Some day, some one, will find out, that recursion is im/possible here. (#define Z_PROBE_END_SCRIPT "G29\nG1 Z10")

[test3012.txt](https://github.com/MarlinFirmware/Marlin/files/175945/test3012.txt)

**EDITED**

AnHardt commented:
```
void get_command() {

  static char serial_line_buffer[MAX_CMD_SIZE];
  static boolean serial_comment_mode = false;

  if (drain_queued_commands_P()) return; // priority is given to non-serial commands
```

That's too late. It's working on the wrong side of the buffer. If there is anything in the buffer now, that will overtake the commands coming from drain_queued_commands_P(). 
It works for Octoprint, Pronterface, RH in pingpong mode, because they are waiting for the "ok" of the **single** line of code we have in the buffer. It does not work for sd printing, RH in prefill mode, because there is **more than one** line used in the buffer.

We have to drain in, or just before, `process_next_command()`. A kind of second Q.

thinkyhead commented:
That makes sense. Perhaps the drain function should be integrated with the rest of the queue handling. I will fix this up asap.

thinkyhead commented:
**[edit] That looks like older code in `get_command`, perhaps…? But here are my comments with the assumption that it is still relevant:**

I'm not entirely sure I understand the problem after all. The intent is to make sure that any multi-line "immediate" commands — those coming from some LCD click options, plus things like `Z_PROBE_END_SCRIPT` — will be processed consecutively with no other commands able to get in-between. (So it now –since I rewrote `drain_queued_commands_P`– returns `true` if it has any command left to add, so the caller can decide to call it again to add more now or later.)

I don't think it is intended to get ahead of – or cancel – commands already in the buffer, if that's what you mean.

Since `drain_queued_commands_P` only adds the next "immediate command" to the queue, and the caller is waiting until the next loop to do the next one, it might seem to be holding things up. We could have it add as many commands as it can instead, something like:

``` cpp
while (drain_queued_commands_P()) if (commands_in_queue >= BUFSIZE) return;
```

On a related note, I realized today that the "host keepalive" feature will be important to mitigate the drain function blocking the other queues, should someone stick "M1" into their `Z_PROBE_END_SCRIPT`, or any other very long operation.

thinkyhead committed:
drain_queued_commands_P now returns true if there are more

AnHardt commented:
At the moment a configured Z_PROBE_END_SCRIPT does nothing.

thinkyhead commented:
Whoops! #3184 reverts `drain_queued_commands_P`. Apparently there's something odd about reading `PSTR` strings…
